src/bobtail/dag.cpp

```diff
@@ -66,46 +61,113 @@ uint16_t CDagNode::GetNodeScore()
-    next_subgraph_id++;
-    for (auto &node : _dag)
+    std::set<COutPoint> new_spends;
+    for (auto &tx : new_node->subblock.vtx)
     {
-        if (tree_ids.count(node->subgraph_id))
+        if (tx->IsProofBase() == false)
         {
-            node->subgraph_id = new_subgraph_id;
+            for (auto &input : tx->vin)
+            {
+                // TODO : change to contains in c++17
+                if (spent_outputs.count(input.prevout) != 0)
+                {
+                    return false;
+                }
+                new_spends.emplace(input.prevout);
+            }
         }
     }
-    return new_subgraph_id;
+    // change to merge in c++17
+    spent_outputs.insert(new_spends.begin(), new_spends.end());
+    _dag.emplace_back(new_node);
+    return true;
 }
 
-void CDagForrest::Clear()
+void CBobtailDagSet::SetNewIds(std::priority_queue<int16_t> &removed_ids)
 {
-    next_subgraph_id = 0;
-    _dag.clear();
+    int16_t last_value;
+    for (auto riter = vdags.rbegin(); riter != vdags.rend(); ++riter)
+    {
+        last_value = removed_ids.top();
+        // TODO : dont use assert here
+        assert(riter->id != last_value);
+        if (riter->id > last_value)
+        {
+            riter->id = riter->id - removed_ids.size();
+        }
+        else // <
+        {
+            removed_ids.pop();
+            riter->id = riter->id - removed_ids.size();
+        }
+        if (removed_ids.empty())
+        {
+            break;
+        }
+    }
+    // do a check to ensure everything lines up
+    for (size_t i = 0; i < vdags.size(); ++i)
+    {
+        // TODO : dont use assert here
+        assert(i == vdags[i].id);
+    }
 }
 
-CDagNode* CDagForrest::Find(const uint256 &hash)
+bool CBobtailDagSet::MergeDags(std::set<int16_t> &tree_ids, int16_t &new_id)
 {
-    // TODO : replace with std::find
-    auto iter = _dag.begin();
-    while (iter != _dag.end())
+    int16_t base_dag_id = *(tree_ids.begin());
+    // remove the first element, it is not being deleted
+    tree_ids.erase(tree_ids.begin());
+    for (auto &id : tree_ids)
     {
-        CDagNode* temp = *iter;
-        if (temp->hash == hash)
+        if (id < 0 || (size_t)id >= vdags.size())
         {
-            return *iter;
+            return false;
         }
-        ++iter;
+        for (CDagNode* node : vdags[id]._dag)
+        {
+            vdags[base_dag_id].Insert(node);
+        }
+    }
+    std::priority_queue<int16_t> removed_ids;
+    // erase after we move all nodes to ensure indexes still align
+    // go in reverse order so indexes still align
+    for (auto riter = tree_ids.rbegin(); riter != tree_ids.rend(); ++riter)
+    {
+        vdags.erase(vdags.begin() + (*riter));
+        removed_ids.push(*riter);
+    }
+    SetNewIds(removed_ids);
+    new_id = base_dag_id;
+    return true;
+}
+
+void CBobtailDagSet::Clear()
+{
+    vdags.clear();
+}
+
+CDagNode* CBobtailDagSet::Find(const uint256 &hash)
```

Comment:
For the Find method, I wonder if we will regret obscuring the DAG where it came from? Perhaps it's better to return a `std::pair<int16_t, CDagNode*>`? I don't have strong feelings about this.

Reply:
is searching for a dag that a node belongs to something you find yourself needing in other areas of the code?

Reply:
No, not presently. Just trying to think ahead. I think we should leave things as you have implemented them for now.